src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -370,126 +418,410 @@ private sealed class SendFileOperation : WriteOperation
+                                    // No more operations to process
+                                    _tail = null;
+                                    _state = QueueState.Ready;
+                                    _sequenceNumber++;
+                                    if (TraceEnabled) Trace(context, $"Exit (finished queue)");
+                                }
+                                else
+                                {
+                                    // Pop current operation and advance to next
+                                    nextOp = _tail.Next = op.Next;
+                                }
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // Check for retry and reset queue state.
+
+                        using (Lock())
+                        {
+                            if (_state == QueueState.Stopped)
+                            {
+                                Debug.Assert(_tail == null);
+                                if (TraceEnabled) Trace(context, $"Exit (stopped)");
+                            }
+                            else
+                            {
+                                Debug.Assert(_state == QueueState.Processing, $"_state={_state} while processing queue!");
+
+                                if (observedSequenceNumber != _sequenceNumber)
+                                {
+                                    // We received another epoll notification since we previously checked it.
+                                    // So, we need to retry the operation.
+                                    Debug.Assert(observedSequenceNumber - _sequenceNumber < 10000, "Very large sequence number increase???");
+                                    observedSequenceNumber = _sequenceNumber;
+                                    nextOp = op;
+                                }
+                                else
+                                {
+                                    _state = QueueState.Waiting;
+                                    if (TraceEnabled) Trace(context, $"Exit (received EAGAIN)");
+                                }
+                            }
+                        }
+                    }
+
+                    if (needCallback || nextOp == null)
+                    {
+                        break;
+                    }
+
+                    op = nextOp;
+                }
+
+                if (needCallback)
                 {
-                    operation.Next = operation;
-                    _tail = operation;
+                    if (nextOp != null)
+                    {
+                        Debug.Assert(_state == QueueState.Processing);
+
+                        // Spawn a new work item to continue processing the queue.
+                        ThreadPool.QueueUserWorkItem(s_processingCallback, context);
```

Comment:
Ah, so we won't actually use the work-stealing queues.  It'd be interesting to implement https://github.com/dotnet/corefx/issues/12442 and then see whether setting preferLocal:true here helps or hurts.  That would likely depend on how fast op.DoCallback is (but it's user code, so it's hard to say) and how loaded the pool is.